File: LeetCode/0264. Ugly Number II/solution.cpp
/**
 * 596 / 596 test cases passed.
 * Runtime: 4 ms
 * Memory Usage: 5.9 MB 
 */
class Solution {
public:
    int nthUglyNumber(int n) {
        int dp[n + 1];
        dp[1] = 1;
        int p2, p3, p5 = p3 = p2 = 1;
        for (int i = 2; i <= n; ++ i) {
            int num2 = dp[p2] * 2;
            int num3 = dp[p3] * 3;
            int num5 = dp[p5] * 5;
            int next = min(num2, min(num3, num5));
            if (next == num2) p2 ++;
            if (next == num3) p3 ++;
            if (next == num5) p5 ++;
            dp[i] = next;
        }
        return dp[n];
    }
};

File: LeetCode/0264. Ugly Number II/solution.py
"""
596 / 596 test cases passed.
Runtime: 128 ms
Memory Usage: 14.6 MB
"""
class Solution:
    def nthUglyNumber(self, n: int) -> int:
        dp = [1] * n
        p2, p3, p5 = 0, 0, 0
        for i in range(1, n):
            n2, n3, n5 = dp[p2] * 2, dp[p3] * 3, dp[p5] * 5
            dp[i] = min(n2, n3, n5)
            if n2 == dp[i]:
                p2 += 1
            if n3 == dp[i]:
                p3 += 1
            if n5 == dp[i]:
                p5 += 1
        return dp[n - 1]
